Allow wildcard in patches search folder path

indykoning opened the pull request:
Closes: #75 

This PR adds the functionality to add wildcards to the path of folders to search in

Example configuration for this would be
```
    "extra": {
        "patcher": {
            "search": ["patches", "app/code/<vendor>/*/patches", "vendor/<vendor>/*/patches"]
        }
    }
```

jooname commented:
Apologies for the (very) delayed followup!

I had a look at this and checked the changes against the package testsuite (with PHP 8.2 and Composer 2.7.0) but it appears this won't work if your patches config is as follows: `"patches-search": "patches"`. It won't crash or anything, it just doesn't seem to find the patches anymore. Granted there's no default value for this and setting the value is up to the users of the package, but anyone having set the config in non-glob pattern will have a surprise coming up if they updated to this change.

I would suggest modifying this a bit: since globbing usually involves wildcard usage, how about checking for existence of wildcard before calling `glob` and then only doing it if there's one involved? Could be like this:

```
        foreach ($source as $sourcePath) {
            $files = [$sourcePath];
            if (strpos($sourcePath, '*') !== false) {
                $files = glob($sourcePath);
            }

            if (!$files) {
                continue;
            }

            foreach ($files as $item) {
                $paths = $this->fileSystemUtils->collectFilePathsRecursively(
                    $basePath . DIRECTORY_SEPARATOR . $item,
                    sprintf('/%s/i', PluginConfig::PATCH_FILE_REGEX_MATCHER)
                );

                $results[] = $this->createPatchDefinitions($basePath, $paths);
            }
        }
```
(Used `strpos` for backwards compatibility)

indykoning commented:
That's an entirely valid usecase i have not used. Nice catch!
Even though we might miss some globs as many combinations are possible i think we'll catch most. I would like to add a check for `?` in there too. The checks for `[...]` or `{a,b,c}` will probably be too heavy.

Due to the added size, should we maybe move this code to be with the fileSystemUtils?

indykoning commented:
I've update the code, moved it to the FileSystemUtils, implemented the [GlobIterator](https://www.php.net/manual/en/class.globiterator.php) instead of calling `glob()`

As an added bonus in glob mode it would support matching files directly now (Though realistically it wouldn't make sense to use it a lot, as collectPathsRecursively already only matches their pattern)

indykoning commented on src/Utils/FileSystemUtils.php:
```diff
@@ -9,13 +9,47 @@ class FileSystemUtils
         });
     }
 
+    public function collectGlobs($rootPath, $pattern)
```
Should i follow te `Recursively` naming schema here?
It does not explicitly use a `RecursiveIterator` like `collectPathsRecursively` does, but it does call `collectPathsRecursively`

jooname commented:
Ran the testsuite with these changes on PHP 7.4 and 8.4, no issues with the new changes! Let's move forward with these, I'll create a minor release.

Thank you for the contribution and once again apologies for taking so long to look into this!
